packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts

```diff
@@ -73,4 +88,111 @@ export class ApplicationExtension<
+     * @returns SerializedExtension - The serialized extension data.
+     * @throws Error if the routes have not been loaded.
+     */
+    public serialize(): SerializedExtension {
+        if (this._cachedRoutes === null) {
+            throw new Error('Routes have not been loaded. Call getRoutes() before serializing')
+        }
+        const serializedRoutes = this._cachedRoutes.map((route) => {
+            if (!route.component?.displayName) {
+                throw new Error(
+                    `Component for route with path "${
+                        route.path
+                    }" is missing a displayName in ${this.getName()} extension`
+                )
+            }
+            return {
+                path: route.path,
+                componentName: route.component?.displayName
+            }
+        })
+        return {
+            routes: serializedRoutes
+        }
+    }
+
+    /**
+     * Returns a map of component names to components that are used to deserialize the extension data.
+     *
+     * @protected
+     * @returns ComponentMap - The map of component names to components.
+     */
+    protected getComponentMap?(): ComponentMap
+
+    /**
+     * Called on the client to deserialize the extension data that was serialized on the server.
+     *
+     * @param serializedExtension - The serialized extension data.
+     * @returns DeserializedExtension - The deserialized extension data.
+     * @throws Error if getComponentMap() is not defined.
+     * @throws Error if the deserialized component cannot be found in the component map.
+     */
+    private deserialize(serializedExtension: SerializedExtension): DeserializedExtension {
+        if (!this.getComponentMap) {
+            throw new Error(
+                `${this.getName()}.getRoutes() is async but does not define getComponentMap()`
+            )
+        }
+
+        const componentMap = this.getComponentMap()
+        const routes = serializedExtension.routes.map(({path, componentName}) => {
+            const component = componentMap[componentName]
+
+            if (!component) {
+                throw new Error(
+                    `${componentName} component could not be deserialized for route with path: ${path}`
+                )
+            }
+
+            return {
+                path,
+                exact: true,
+                component
+            }
+        })
+        return {routes}
+    }
+
+    private handleAsyncRoutes() {
+        if (!isServerSide) {
+            // Deserialize the routes on the client to ensure the latest routes are loaded on the client
+            this._cachedRoutes = this.deserialize(window.__EXTENSIONS__[this.getName()]).routes
+        }
+
+        // Apply caching for the getRoutes method
+        this.applyCacheForMethod('getRoutes', '_cachedRoutes')
+    }
+
+    private isGetRoutesAsync(): boolean {
+        const routes = this.getRoutes()
+        return routes instanceof Promise
+    }
```

Comment:
Yeah.. so I guess I didn't foresee this being the case. But as you have done the work it's more obvious that we can't tell what the function is returning unless we actually call it. 

And it feels weird to have this async call happen when we are just checking, even tho we have caching, I think its probably not the cleanest solution. 

Although there are other ways to do this, they aren't compatible with our code because the react-sdk is in javascript.

I hate to say this, but I think we'll have to go back to the `getRoutes/getRoutesAsync` solution.